File: FileService/src/main/java/net/blueshell/fileservice/FileController.java
package net.blueshell.fileservice;

import net.blueshell.common.dto.FileDTO;
import net.blueshell.db.BaseController;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;

@Controller
public class FileController extends BaseController<FileService, FileRepository> {

    private final FileMapper fileMapper;

    @Autowired
    public FileController(FileService service, FileRepository repository, FileMapper fileMapper) {
        super(service, repository);
        this.fileMapper = fileMapper;
    }

    @GetMapping("/files/{filename:.+}")
    @ResponseBody
    public ResponseEntity<Resource> downloadFile(@PathVariable String filename) {
        File file = service.findByName(filename);
        return service.prepareFileResponse(file);
    }

    @PostMapping("/files")
    public List<FileDTO> uploadFile(List<FileDTO> dtos) {
        List<File> files = fileMapper.fromDTOs(dtos);
        return fileMapper.toDTOs(files);
    }
}

File: api/src/main/kotlin/net/blueshell/api/file/web/FileController.kt
package net.blueshell.api.file.web

import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.constraints.NotNull
import net.blueshell.api.file.application.FileService
import net.blueshell.api.file.web.dto.FileDTO
import net.blueshell.api.file.web.mapping.asDto
import net.blueshell.api.shared.enums.FileType
import net.blueshell.api.shared.web.BaseController
import org.springframework.core.io.Resource
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile

@RestController
@Tag(name = "Files")
class FileController(service: FileService) : BaseController<FileService>(service) {
    @GetMapping("/events/{eventId}/banners")
    @PreAuthorize("hasAuthority('BOARD') || hasPermission(#eventId, 'Event', 'read')")
    fun downloadEventBanner(@PathVariable eventId: Long): ResponseEntity<Resource> {
        val file = service.findByBannerEventId(eventId)
        return service.prepareFileResponse(file)
    }

    @PostMapping(value = ["/events/banners"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    @ResponseStatus(
        HttpStatus.CREATED
    )
    @PreAuthorize("hasAuthority('COMMITTEE')")
    fun uploadEventBanner(
        @RequestPart("file") @NotNull(message = "File is required") @net.blueshell.api.file.web.validation.FileSize(max = 2 * 1024 * 1024) @net.blueshell.api.file.web.validation.AllowedContentTypes(
            "image/png",
            "image/jpeg",
            "image/jpg",
            "image/webp",
            "image/gif"
        ) file: @NotNull(message = "File is required") MultipartFile
    ): FileDTO {
        val stored = service.storeMultipart(file, FileType.EVENT_BANNER)
        return stored.asDto()
    }
}
